File: linked-lists/implementations/python/singly_linked_list.py
from typing import Optional, Union, Generic
from linked_list_nodes import T, Node


class LinkedList(Generic[T]):
    """
    Represents a singly linked list data structure.

    A singly linked list is a linear collection of nodes, where each node contains a value and a reference
    to the next node in the sequence. The list is traversed by following the next pointers from one node
    to the next.

    Operations.
    - Insertion at Head/Beginning
    - Insertion at Tail
    - Traversal
    - Delete at Head
    - Delete at Tail
    - Insert at a given Position
    - Delete at a given Position
    - Search
    """

    def __init__(self, head:Optional[Node]=None) -> None:
        """
        Initializes a singly linked list with the given node as the head.

        Args:
            head (Optional[Node], optional): The first node in the list. Defaults to None if not provided, and 
                the list is initialized as empty.
        """
        self.head = head


    def prepend(self, val:T)->None:
        """
        Inserts a node at the beginning. The node is treated as the head.

        Args:
            val (T): Represents the value/data to be stored in the node.
        """
        new_node = Node(val, self.head)
        self.head = new_node


    def append(self, val:T)->None:
        """
        Adds a new node to the end of the of the list.

        Args:
            val (T): Represents the value/data to be stored in the node.
        """
        current = self.head
        new_node = Node(val)

        # Handle edge case, if list is empty set the new node as the head
        if self.head is None:
            self.head = new_node
            return

        # Traverse to the end of the list
        while current.next:
            current = current.next

        current.next = new_node

    def print_list(self)->None:
        """
        Prints the values in the linked list, helping with understanding traversal.

        Raises:
            Exception: If list is empty.

        Example output:
            0->1->2->3->None
        """

        if self.head is None:
            raise ValueError("Linked list is empty")
            

        current = self.head

        while current:
            print(current.val, end='->')
            current = current.next

        print("None")

    def search(self, val:T, return_index:bool=False)->Union[bool, T]:

             
        """
        Finds the first occurence of a node by given value.

        Args:
            val (T): Value of the node to be searched.
            return_index (bool, optional): Whether to return the index instead of a boolean.


        Raises:
            ValueError: If the list is empty

        Returns:
            Union[bool, T]: Boolean indicating presence or index of the value
        """

        # Handle edge case: if the list is empty
        if self.head is None:
            raise ValueError("Linked list is empty")

        # If the node we're looking for is the head
        if self.head.val == val:
            return 0 if return_index else True

        # Traverse through list to find the node
        current = self.head
        index = 0

        while current:
            index+=1
            if current.val == val:
                return index if return_index else True
            current = current.next

        return -1 if return_index else False


    def delete(self, val:T)->None:
        """
         Deletes the first occurrence of a node with the given value.

        Args:
            val (T): Value of the node to be deleted.

        Raises:
            ValueError: If the list is empty or the node is not found.
        """

        # Handle edge case: if the list is empty
        if self.head is None:
            raise ValueError("Cannot complete delete operation. Linked list is empty!") 

        # If the node to be deleted is head
        if self.head.val == val:
            self.head = self.head.next
            print(f'Node ({val}) deleted successfully.')
            return

        # Traverse the list to find the node to delete
        current = self.head
        while current.next:
            if current.next.val == val:
                current.next = current.next.next
                print(f'Node ({val}) deleted successfully.')
                return

            current = current.next
        # Node not found
        raise ValueError(f'Node ({val}) Not found!')
        
    def insert_at(self, val:T, pos:T)->None:
        """
        Adds a node at a given position or index.

        Args:
            val (T): Represents the value/data to be stored in the node.
            pos (T): Represents the index/position where the node should be inserted.

        Raises:
            ValueError: If position is invalid/out of range or list is empty.
        """
        new_node  = Node(val)

        # Handle edge case: negative index/pos
        if pos < 0:
            raise ValueError("Index or posistion cannot be negative")
      
        # Handle edge case: if List is empty and pos is greater than 0
        if self.head is None:
            raise ValueError("Cannot insert at position greater than 0 in an empty list")

        # Handle edge case: insert at the beginning, when pos=0
        if pos == 0:
            new_node.next = self.head
            self.head= new_node

        # Otherwise, Traverse to find the node before the pos/index
        current = self.head
        count = 0

        while current and count < pos-1:
            current = current.next
            count+=1

        # Handle edge case: check if index is out of range and current is none
        if current is None:
            raise ValueError(f"Cannot insert at position {pos}. Index is out of range!")

        new_node.next = current.next
        current.next = new_node

    def delete_at(self,  pos:T)->None:
        """
        Deletes a node at a given position/index.

        Args:
            pos (T): Represents the index/position of the node to be deleted.

        Raises:
            ValueError: If position is invalid/out of range or list is empty.
        """
        # Handle edge case: negative index/pos
        if pos < 0:
            raise ValueError("Index or posistion cannot be negative.")

        
        # Handle edge case: if List is empty and pos is greater than 0
        if self.head is None:
            raise ValueError("Cannot complete delete operation. Linked list is empty!")

        # Handle edge case: delete at the beginning, when pos=0,
        if pos == 0:
            self.head=self.head.next
            return

        # Traverse the list to find the node before the index
        current = self.head
        count = 0

        while current and count < pos-1:
            current = current.next
            count +=1
        
        # Handle edge case: check if index is out of range and current is none
        if current is None or current.next is None:
            raise ValueError(f"Cannot delete Node at position {pos}. Index is out of range!")

        current.next = current.next.next

        
            

    

        

File: linked-lists/implementations/typescript/singly_linked_list.ts
import { SinglyLinkedNode } from "./linked_list_nodes";
//  ----------------------------------------------------------------------

/**
 *  Represents a singly linked list data structure.
 *  A singly linked list is a linear collection of nodes, where each node contains a value and a reference
    to the next node in the sequence. The list is traversed by following the next pointers from one node
    to the next.
    
*   Operations.
    - Insertion at Head/Beginning
    - Insertion at Tail
    - Traversal
    - Delete at Head
    - Delete at Tail
    - Insert at a given Position
    - Delete at a given Position
    - Search
 */
export default class SinglyLinkedList<T> {
  head: SinglyLinkedNode<T> | null;

  /**
   * Initializes a singly linked list with the given node as the head.
   * @param head: The first node in the list. Defaults to null if not provided, and 
        the list is initialized as empty.
   */
  constructor(head: SinglyLinkedNode<T> | null = null) {
    this.head = head;
  }

  /**
   * Adds a new node to the end of the of the list.
   * @param value: Represents the value/data to be stored in the node.
   */
  append(value: T): void {
    const newNode = new SinglyLinkedNode(value);

    // Handle edge case, if list is empty set the new node as the head
    if (this.head === null) {
      this.head = newNode;
      return;
    }

    // Traverse to the end of the list
    let current = this.head;

    while (current.next) {
      current = current.next;
    }

    current.next = newNode;
  }

  /**
   * Inserts a node at the beginning. The node is treated as the head.
   * @param value: Represents the value/data to be stored in the node.
   */
  prepend(value: T): void {
    const newNode = new SinglyLinkedNode(value);

    newNode.next = this.head;
    this.head = newNode;
  }

  /**
   * Prints the values in the linked list, helping with understanding traversal.
   */
  printList(): void {
    // Handle edge case, if list is empty set the new node as the head
    if (this.head === null) {
      throw new Error("Linked list is empty");
    }

    let current: SinglyLinkedNode<T> | null = this.head;
    let values: T[] = [];

    while (current) {
      values.push(current.value);
      current = current.next;
    }

    console.log(values.join("->"));
  }

  /**
   * Finds the first occurence of a node by given value.
   * @param value: Value of the node to be searched.
   * @param returnIndex): Whether to return the index instead of a boolean.
   * @returns: Boolean indicating presence or index of the value
   */
  find(value: T, returnIndex: boolean = false): boolean | number {
    // Handle edge case, if list is empty set the new node as the head
    if (this.head === null) {
      throw new Error("Linked list is empty");
    }

    // If the node we're looking for is the head
    if (this.head.value === value) {
      return returnIndex ? 0 : true;
    }

    // Traverse through the list to find the node
    let current: SinglyLinkedNode<T> | null = this.head;
    let count: number = 0;

    while (current) {
      if (current.value === value) {
        return returnIndex ? count : true;
      }
      current = current.next;
      count++;
    }

    return returnIndex ? -1 : false;
  }

  /**
   * Deletes the first occurrence of a node with the given value.
   * @param value: Value of the node to be deleted.
   * @returns
   */
  delete(value: T): void {
    // Handle edge case, if list is empty set the new node as the head
    if (this.head === null) {
      throw new Error("Linked list is empty");
    }

    // If the node we're looking for is the head
    if (this.head.value === value) {
      this.head = this.head.next;
      return;
    }

    // Traverse to find the node to delete
    let current: SinglyLinkedNode<T> | null = this.head;

    while (current) {
      if (current.next?.value === value) {
        current.next = current.next.next;
        return;
      }
      current = current.next;
    }

    throw new Error(`Node ${value} Not found!`);
  }

  /**
   * Adds a node at a given position or index.
   * @param value: Represents the value/data to be stored in the node.
   * @param index: Represents the index/position where the node should be inserted.
   */
  insert_at(value: T, index: number): void {
    const newNode = new SinglyLinkedNode<T>(value);
    // Handle edge case: negative index
    if (index < 0) {
      throw new Error("Index or posistion cannot be negative");
    }

    // Handle edge case: if List is empty and index is greater than 0
    if (this.head === null && index > 0) {
      throw new Error(
        "Cannot insert at position greater than 0 in an empty list"
      );
    }

    // Handle edge case: insert at the beginning, when index=0
    if (index === 0) {
      newNode.next = this.head;
      this.head = newNode;
      return;
    }

    // Otherwise, Traverse to find the node before the index
    let current = this.head;
    let count = 0;

    while (current && count < index - 1) {
      current = current.next;
      count++;
    }

    // Handle edge case: check if index is out of range and current is null
    if (current === null) {
      throw new Error(
        `Cannot insert at position ${index}. Index is out of range!`
      );
    }

    // insert the node
    newNode.next = current.next;
    current.next = newNode;
  }

  /**
   * Deletes a node at a given position/index.
   * @param index: Represents the index/position of the node to be deleted.
   */
  delete_at(index: number): void {
    // Handle edge case: negative index
    if (index < 0) {
      throw new Error("Index or posistion cannot be negative");
    }

    // Handle edge case: if List is empty and index is greater than 0
    if (this.head === null) {
      throw new Error(
        "Cannot complete delete operation. Linked list is empty!"
      );
    }

    // Handle edge case: delete at the beginning, when index=0
    if (index === 0) {
      this.head = this.head?.next;
      return;
    }

    // Otherwise, Traverse to find the node before the index
    let current: SinglyLinkedNode<T> | null = this.head;
    let count = 0;

    while (current && count < index - 1) {
      current = current.next;
      count++;
    }

    // Handle edge case: check if index is out of range and current is null
    if (current === null || current?.next === null) {
      throw new Error(
        `Cannot delete Node at position ${index}. Index is out of range!`
      );
    }

    current.next = current.next.next;
  }
}
